### PR TITLE
Make InternalExclusiveModeGpuDiscoveryPlugin and ExplainPlanImpl as protected class.

### PR DESCRIPTION
**Pull request by nartal1**
This fixes https://github.com/NVIDIA/spark-rapids/issues/4022.  
Went through the docs and checked all the public classes which can be set. In this PR, we are making InternalExclusiveModeGpuDiscoveryPlugin as protected as it should be loaded only through reflection. 
Updated the comments for  ParquetCachedBatchSerializer as well. 
Rest of the public classes - GpuKryoRegistrator and RapidsShuffleManager doesn't require any changes.

### CODE REVIEW

**Comment by gerashegalov** [comment]
from the classes referenced in ShimLoader via newInstanceOf it looks we need to cover `com.nvidia.spark.rapids.ExplainPlanImpl` as well.

**Comment by nartal1** [comment]
> from the classes referenced in ShimLoader via newInstanceOf it looks we need to cover `com.nvidia.spark.rapids.ExplainPlanImpl` as well.

Thanks, @gerashegalov  for pointing this out. Updated the PR.  PTAL.

File: sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala [hunk 1]
```diff
@@ -257,7 +257,8 @@ private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) e
 }
 
 /**
- * This class assumes, the data is Columnar and the plugin is on
+ * This class assumes, the data is Columnar and the plugin is on. This is meant to be loaded
+ * by reflection only.
```
Comment (gerashegalov) [inline sql-plugin/src/main/311+-all/scala/com/nvidia/spark/rapids/shims/v2/ParquetCachedBatchSerializer.scala:261]: last nit: let us make this and other comments more precise: 
> Note, this class should not referenced directly in source code. It should be loaded by reflection using ShimLoader.newInstanceOf, see  ./docs/dev/shims.md

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by gerashegalov** [comment]
build

**Comment by nartal1** [comment]
Thanks @gerashegalov  for the review.

**Comment by nartal1** [comment]
build

**Pull request merged**